server/mediaserver.go

```diff
@@ -1037,6 +1038,83 @@ func getPlaylistsFromStore(ctx context.Context, sess drivers.OSSession, manifest
 	return filesMap, jsonFiles, latestPlaylistTime, nil
 }
 
+func (s *LivepeerServer) streamMP4(w http.ResponseWriter, r *http.Request, jpl *core.JsonPlaylist, track string) {
+	w.Header().Set("Access-Control-Allow-Origin", "*")
+	contentType, _ := common.TypeByExtension(".mp4")
+	w.Header().Set("Content-Type", contentType)
+	w.Header().Set("Connection", "keep-alive")
+
+	tc := ffmpeg.NewTranscoder()
+	or, ow, err := os.Pipe()
+	if err != nil {
+		glog.Error(err)
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	defer func() {
+		tc.StopTranscoder()
+		ow.Close()
+	}()
+	go io.Copy(w, or)
+	oname := fmt.Sprintf("pipe:%d", ow.Fd())
+	out := []ffmpeg.TranscodeOptions{
+		{
+			Oname: oname,
+			VideoEncoder: ffmpeg.ComponentOptions{
+				Name: "copy",
+			},
+			AudioEncoder: ffmpeg.ComponentOptions{
+				Name: "copy",
+			},
+			Profile: ffmpeg.VideoProfile{Format: ffmpeg.FormatNone},
+			Muxer: ffmpeg.ComponentOptions{
+				Name: "mp4",
+				Opts: map[string]string{"movflags": "frag_keyframe+negative_cts_offsets+omit_tfhd_offset+disable_chpl+default_base_moof"},
+			},
+		},
+	}
+	for _, seg := range jpl.Segments[track] {
+		if seg.GetDiscontinuity() {
+			tc.Discontinuity()
+		}
+
+		ir, iw, err := os.Pipe()
```

Comment:
Instead of creating a new pipe, writing the downloaded segment data into the pipe, calling `Transcode()` and closing the pipe for each segment, could it make sense to create a single input pipe, write all downloaded segment data into the pipe and call `Transcode()` once? Something like:

```golang
ir, iw, err := os.Pipe()
if err != nil { ... }

defer func() {
    tc.StopTranscoder()
    ow.Close()
    iw.Close()
}

for _, seg := range jpl.Segments[track] {
    ...
    go func() {
        // Download segment data
       _, err = io.Copy(iw, resp.Body)
       if err != nil { ... }
    }
}

fname := fmt.Sprintf("pipe:%d", ir.Fd())
in := &ffmpeg.TranscodeOptions{Fname: fname, Transmuxing: true}
_, err = tc.Transcode(in, out)
if err != nil { ... }
```

If this works, the main benefit I see of this approach is a slightly cleaner separation of logic within the function i.e. for loop responsible for just downloading/writing segment data into a single pipe and then everything else is handled by the input/output pipe + Transcode() call.

Reply:
No, because we need to signal discontinuity to transcoding loop. 
Say, user streamed for 10 min, stopped, and then immediately started again and streamed for another 10 min.
We will have this as single recording, but timestamps of the second parted will be started from zero.
To make from that one mp4 file we need to shift timestamps of second part to make on continuous stream.
To do that we need a way to signal that discontinuity to the transcoding loop - which is not possible if have one loop for the whole transmuxing process.
That's why I've changed lpms so for transmuxing it closes input after Transcode call, but keeps open output.
That way I can signal discontiniuty to the transmuxer between Transcode calls.
 https://github.com/livepeer/go-livepeer/blob/d87fcc6006af3684c38947a9fd7b831315b703e3/server/mediaserver.go#L1079-L1081